### PR TITLE
Support custom spec sources

### PR DESCRIPTION
**Pull request by MikailBag**
Usage example:
```bash
kubectl proxy &
cd k8s-openapi-codegen
VERSIONS=v1.20=http://127.0.0.1:8001/openapi/v2 cargo run
```
This change is not useful on its own, but it eventually it should allow support non-builtin API groups, such as `metrics.k8s.io`.

I have not written any documentation yet. What is the best location for docs?

To summarize, using the `VERSIONS` environment variable user can:
1) Request that only a subset of kubernetes versions should be handled
2) For each version they can provide custom specification URL
E.g.:
`VERSIONS=v.19,v1.20=http://foo.bar/spec.json`
Here codegen will generate v1.19 (using spec from k/k github repository) and v1.20 (using spec from `http://foo.bar/spec.json`)

### CODE REVIEW

**Comment by Arnavion** [comment]
>I have not written any documentation yet. What is the best location for docs?

Well `-codegen` has been an "internal" tool up to this point, with the other crates being the point of the repo. But we can change that of course.

Just add it to `k8s-openapi-codegen/README.md`

>To summarize, using the `VERSIONS` environment variable user can:

It would be better to take it as CLI args:

```sh
# Generate all with built-in URLs
k8s-openapi-codegen

# Only generate 1.19 and 1.20 with built-in URLs
k8s-openapi-codegen --generate 1.19 --generate 1.20

# Only generate 1.19 and 1.20, and override URL for 1.20
k8s-openapi-codegen --generate 1.19 --generate 1.20:http://foo.bar/spec.json
```

(Use `structopt` please.)

File: k8s-openapi-codegen/src/main.rs [hunk 1]
```diff
@@ -41,6 +41,35 @@ impl std::fmt::Debug for Error {
...
+}
+
+fn parse_requested_versions(spec: &str) -> Vec<RequestedVersion> {
+	let mut res = Vec::new();
+	for item in spec.split(',') {
+		let mut it = item.split('=');
+		let name = it.next().expect("empty requested version");
+		let url = it.next();
+		let mut found_version = false;
+		for ver in supported_version::ALL.iter().copied() {
```
Comment (Arnavion) [inline k8s-openapi-codegen/src/main.rs:56]:
```rust
if let Some(&ver) = supported_version::ALL.iter().find(|&ver| ver.name() == name) {
```

File: k8s-openapi-codegen/src/main.rs [hunk 2]
```diff
@@ -41,6 +41,35 @@ impl std::fmt::Debug for Error {
...
+				res.push(RequestedVersion {
+					ver,
+					overriden_spec_url: url.map(ToString::to_string)
+				});
+				found_version = true;
+				break;
+			}
+		} 
+		if !found_version {
+			panic!("unknown requested version");
```
Comment (Arnavion) [inline k8s-openapi-codegen/src/main.rs:67]: Return an `Err()` like the rest of the code, and include `{:?}` of `found_version` in the message.

File: k8s-openapi-codegen/src/main.rs [hunk 3]
```diff
@@ -99,7 +140,7 @@ fn main() -> Result<(), Error> {
 	result
 }
 
-fn run(supported_version: supported_version::SupportedVersion, out_dir_base: &std::path::Path, client: &reqwest::blocking::Client) -> Result<(), Error> {
+fn run(supported_version: supported_version::SupportedVersion, out_dir_base: &std::path::Path, client: &reqwest::blocking::Client, spec_url_override: Option<&str>) -> Result<(), Error> {
```
Comment (Arnavion) [inline k8s-openapi-codegen/src/main.rs:143]: Just take `spec_url: &str` and make the caller specify either the default from `SupportedVersion::spec_url` or the user one.

**Review by Arnavion — CHANGES_REQUESTED** [review]

**Comment by Arnavion** [comment]
I fixed a few things and pushed to https://github.com/Arnavion/k8s-openapi/compare/pr/90

- Some incorrect whitespace - spaces indentation, trailing whitespace.

- The `--version` parsing was broken because it expected the parameter to contain the `v` prefix. Since `--version=v1.20` would look redundant, I changed the `SupportedVersion::name()` function to not have the `v` prefix.

- The `--version` parsing was broken because it used `split(':')` instead of `splitn(2, ':')`, so URLs would truncate after the protocol and be invalid.

- Tweaked the helptext and error message to be more explanatory.

Does that branch look okay to you?

**Comment by MikailBag** [comment]
> Does that branch look okay to you?

Yes, sure. Thanks for your review and fixes.

**Pull request closed without merging**